Storage/SQL/SQLiteHistory.swift

```diff
@@ -915,49 +915,67 @@ extension SQLiteHistory: SyncableHistory {
         //
-        // That's the greatest-N-per-group problem in SQL. Please read and understand the solution
-        // to this (particularly how the LEFT OUTER JOIN/HAVING clause works) before changing this query!
+        // That's the greatest-N-per-group problem. We used to do this in SQL, joining
+        // the visits table to a subselect of visits table, however, ran into OOM issues (Bug 1417034)
+        //
+        // Now, we want a more dumb approach with no joins in SQL and doing group-by-site and limit-to-N in swift.
+        //
+        // We do this in a single query, rather than the N+1 that desktop takes.
         //
-        // We can do this in a single query, rather than the N+1 that desktop takes.
         // We then need to flatten the cursor. We do that by collecting
         // places as a side-effect of the factory, producing visits as a result, and merging in memory.
 
         // Turn our lazy collection of integers into a comma-seperated string for the IN clause.
         let historyIDs = Array(places.keys)
-        let inClause = "siteID IN ( \(historyIDs.map(String.init).joined(separator: ",")) )"
 
         let sql =
-        "SELECT v1.siteID AS siteID, v1.date AS visitDate, v1.type AS visitType " +
-        "FROM (" +
-        "   SELECT * FROM \(TableVisits) WHERE \(inClause) AND type <> 0" +
-        ") AS v1 " +
-        "LEFT OUTER JOIN \(TableVisits) AS v2 ON v1.siteID = v2.siteID AND v1.date < v2.date " +
-        "GROUP BY v1.date " +
-        "HAVING COUNT(*) < ? " +
-        "ORDER BY v1.siteID, v1.date DESC"
-
-        // Seed our accumulator with empty lists since we already know which IDs we will be fetching.
-        var visits = [Int: [Visit]]()
-        historyIDs.forEach { visits[$0] = [] }
-
-        // Add each visit to its history item's list.
-        let visitsAccumulator: (SDRow) -> Void = { row in
-            let date = row.getTimestamp("visitDate")!
-            let type = VisitType(rawValue: row["visitType"] as! Int)!
-            let visit = Visit(date: date, type: type)
-            let id = row["siteID"] as! Int
-            visits[id]?.append(visit)
-        }
+                "SELECT siteID, date AS visitDate, type AS visitType " +
+                "FROM \(TableVisits) " +
+                "WHERE siteID IN (\(historyIDs.map(String.init).joined(separator: ","))) " +
+                "ORDER BY date DESC"
+
+        // We want to get a tuple Visit and Place here. We can either have an explicit tuple factory
+        // or we use an identity function, and make do without creating extra data structures.
+        // Since we have a known Out Of Memory issue here, let's avoid extra data structures.
+        let rowIdentity: (SDRow) -> SDRow = { $0 }
+
+        // We'll need to runQueryUnsafe so we get a LiveCursor, i.e. we don't get the cursor
+        // contents into memory all at once.
+        let args: Args = []
```

Comment:
Do we need `args`? Can't we just pass `nil`?

Reply:
:+1: